doc/ksitool.1
.TH KSITOOL 1



.SH NAME
.B ksitool
- Guardtime command-line tool to access KSI service.


.SH SYNOPSIS
Sign data file, stream (stdin) or hash:
.br
\fBksitool -s -f\fI file\fR | [\fB-F\fI hash\fR]\fB -o\fI out.ksig\fR [\fImore options\fR]
.br

Extend signature:
.br
\fBksitool -x -i\fI sig.ksig\fB -o\fI ext.ksig\fR [\fB-T\fR] [\fImore options\fR]
.br

Download and verify publication file:
.br
\fBksitool -p -o \fI pubfile\fB --cnstr \fIoid=value \fR... [\fImore options\fR]
.br

Dump publication file:
.br
\fBksitool -p -d -P\fI url\fR | [\fB-b \fIpubfile\fR]\fR
.br

Download publication file:
.br
\fBksitool -p -o \fI pubfile\fB --cnstr \fIoid=value \fR... [\fImore options\fR]
.br

Create publication string:
.br
\fBksitool -p -T \fR[\fImore options]
.br

Verify signature:
.br
\fBksitool -v -i\fI sig.ksig\fR [\fB-f\fI file\fR | \fB-F\fI hash\fR] [\fB-x\fR | \fB-b \fI pubfile\fR] [\fI more options\fR]
.br

Verify signature with publications string:
.br
\fBksitool -v -i\fI sig.ksig\fR \fB--ref\fR \fIpubstr\fR [\fB-f\fI file\fR | \fB-F\fI hash\fR] [\fB-x\fR] [\fI more options\fR]
.br

Verify publication file:
.br
\fBksitool -v -b\fI pubfile\fB --cnstr \fIoid=value \fR... [\fImore options\fR]
.br
.\TODO: uncomment if implemented
.\Display current aggregation root hash value and time:
.\.br
.\.B ksitool --aggre --htime
.\.br
.\
.\Set system time from current aggregation:
.\.br
.\.B ksitool --aggre --setsystime
.\.br


.SH DESCRIPTION

This is a general signing and signature verification tool for Guardtime Keyless Signature services.


.SH OPTIONS

.TP
.B Tasks:
.\TODO: uncomment if implemented
.\.TP
.\.B --aggre
.\Use for getting root hash and time from aggregator.
.TP
.B -p
Use for downloading publication file and creating publication string.
.TP
.B -s, --sign
Use for signing document, stream (stdin) or hash.
.TP
.B -v, --verify
Use for signature and publication file verification.
.TP
.B -x, --extend
Use for online verification or signature extending (permit extending when verifying signature with publication string).

.TP
.B Input/output:
.TP
.BI -b\  file
Use specified publication file.
.TP
.BI -f\  file
File to be signed or verified. \fIfile\fR can be '\fB-\fR' to sign a stream (stdin).
.TP
.BI -F\  hash
Data hash to be signed or verified. Hash format: <alg>:<hash in hex>.
.TP
.BI -H\  alg
Hash algorithm used to hash the file to be signed. Use when signing file (
.B -s -f
).
.TP
.BI -i\  file
Input signature token file to be extended or verified.
.TP
.BI -o\  file
Output file name to store signature token or publication file.
.TP
\fB-D, --dataout\fI  file
Output file name to store data from stream (stdin).
.TP
.BI --ref\  str
Publication string as base 32 encoded string. Use with \fB-v\fR, add \fB-x\fR to permit extending.
.TP
.BI -T\  int
.br
Specify a publication time to extend to (use with \fB-x\fR) or a time to create a new publication code for (use with \fB-p\fR) as the number of seconds since 1970-01-01 00:00:00 UTC or time string formatted as "YYYY-MM-DD hh:mm:ss".
.br

.TP
.B Details:
.TP
.B -d
Print detailed information.
.TP
.BI --log\  file
Dump KSI log into file.
.TP
.B -n
Print signer Name (identity).
.TP
.B --nowarn
Silence warning messages.
.br
.TP
.B -r
Print publication References (use with -vx).
.TP
.B --silent
Silence info and warning messages.
.br
.TP
.B -t
Print service Timing in ms.
.br

.TP
.B Configuration:
.TP
.BI -c\  int
Network transfer timeout in seconds, after successful Connect.
.TP
.BI -C\  int
Network Connect timeout in seconds (is not supported with tcp client).
.TP
.BI --cnstr\  oid=value
use OID and its expected value to verify publications file PKI signature. At least one constraint must be defined to be able to verify publications file but it is possible to define more. If value part contains spaces use " " to wrap its contents. For common OID's there are string representations: '\fBemail\fR' for 1.2.840.113549.1.9.1, '\fBcountry\fR' for 2.5.4.6, '\fBorg\fR' for 2.5.4.10 and '\fBcommon_name\fR' for 2.5.4.3. Example \fB--cnstr \fI2.5.4.6=EE \fB--cnstr \fIorg="Guardtime AS"\fR.
.TP
.BI --inc\  file
Use configuration file containing command-line parameters. Parameter must be written line by line.
.TP
.BI --pass\  str
Password for authentication.
.TP
.BI -P\  url
Specify publication file URL.
.TP
.BI -S\  url
Specify Signing service URL.
.TP
.BI --user\  str
User name for authentication.
.TP
.BI -V\  file
Use specified OpenSSL-style trust store file for publication file verification. Can have multiple values (-V <file 1> -V <file 2>).
.TP
.BI -W\  dir
Use specified OpenSSL-style trust store directory for publication file verification.
.TP
.BI -X\  url
Specify verification (eXtending) service URL.
.br

.TP
.B Help:
.TP
.B -h, --help
print ksitool help.
.br			


.SH EXIT STATUS

.TP
.B 0
Exit success. Returned if everything is OK.
.br
.TP
.B 1
Exit failure. A general failure occurred.
.br
.TP
.B 3
Invalid command-line parameter. The content or format of command-line parameter is invalid. Also a parameter may be missing.
.br
.TP
.B 4
Invalid format. Input data to KSI library is invalid, for example signature or publication file format is invalid.
.br
.TP
.B 5
Network error. Ksitool is unable to connect to the service, connection is timed out or an HTTP error was returned from the service url.
.br
.TP
.B 6
Verification error. Verification of signature or document hash failed.
.br
.TP
.B 7
Extending error. Error in extending a signature or an error was returned by extender.
.br
.TP
.B 8
Aggregation error. Error returned by aggregator.
.br	
.TP
.B 9
Input / output error. Unable to write or read file.
.br
.TP
.B 10
Cryptographic error. Error may be generated due to untrusted or unavailable hash algorithm or by an invalid PKI signature or untrusted certificate.
.br
.TP
.B 11
HMAC error. HMAC of aggregation or extend response does not match. 
.br
.TP
.B 12
No privileges. Operating system did not grant privileges to perform an operation.
.br
.TP
.B 13
System out of memory.
.br
.TP
.B 14
Authentication error. Aggregation or extending service did not accept user identification parameters.
.br


.SH EXAMPLES

In the following examples it is assumed that default service urls are defined as environment variables. Read example 1 to learn how to define service urls.

\fB1\fR To use ksitool, service urls must be specified. It can be done via environment variables, command-line parameters or a configuration file.

\fB1.1\fR To define default urls as environment variables, KSI_AGGREGATOR, KSI_EXTENDER and KSI_PUBFILE must be defined as shown below: 

.RS
.br
\fBKSI_AGGREGATOR\fR=\fIurl=http://test.com:3333/gt-signingservice pass=test_pass user=test_user\fR
.br
\fBKSI_EXTENDER\fR=\fIurl=http://test.com:8010/gt-extendingservice pass=test_pass user=test_user\fR
.br
\fBKSI_PUBFILE\fR=\fIurl=http://verify.guardtime.com/ksi-publications.bin 1.2.840.113549.1.9.1=publications@guardtime.com 2.5.4.10="Symantec Corporation"\fR
.RE


\fB1.2\fR To define service urls on command-line or via configuration file, following parameters must be defined:
.RS
.br
\fB-X\fI http://test.com:8010/gt-extendingservice
.br
\fB-S\fI http://test.com:3333/gt-signingservice
.br
\fB--user\fI test_user
.br
\fB--pass\fI test_pass
.RE
.br

\fB1.3\fR To use a configuration file, parameters must be written on separate lines, into a file, as in the example above. Configuration file \fIconf\fR  must be included using option:
.br
.RS
\fB--inc\fI conf  
.RE
.br

\fB2\fR To sign a file \fIfile\fR and save signature to \fIsig.ksig\fR call:
.RS
.br
\fBksitool -s -f \fIfile\fR \fB-o \fIsig.ksig\fR
.RE

\fB2.1\fR To sign a stream (stdin), save data from stream to \fIfile\fR and save signature to \fIsig.ksig\fR call:
.RS
.br
\fBksitool -s -f - -D \fIfile\fB -o \fIsig.ksig\fR
.RE

\fB2.2\fR To sign a stream (stdin) and save signature to \fIsig.ksig\fR without saving data from stream call:
.RS
.br
\fBksitool -s -f - -o \fIsig.ksig\fR
.RE

\fB3\fR To sign a data hash (hashed with SHA256) and save the resulting signature to file \fIsig.ksig\fR call:
.RS
.br
\fBksitool -s -o \fIsig.ksig\fB -F\fI SHA-256:c8ef6d57ac28d1b4e95a513959f5fcdd0688380a43d601a5ace1d2e96884690a\fR
.RE

\fB4\fR To sign a data file \fIfile\fR with non-default algorithm \fI SHA1\fR call:
.br
.RS
\fBksitool -s -f \fIfile\fB -H\fI SHA1\fB -o\fI sig.ksig\fR 
.RE

\fB5\fR To verify a signature \fIsig.ksig\fR and file \fIfile\fR it belongs to call:
.RS
\fBksitool -v -i \fIsig.ksig\fB -f\fI file\fR
.RE

\fB6\fR To verify a signature \fIsig.ksig\fR and hash it belongs to call:
.RS
\fBksitool -v -i\fI sig.ksig \fB-F\fI SHA-256:c8ef6d57ac28d1b4e95a513959f5fcdd0688380a43d601a5ace1d2e96884690a\fR
.RE

\fB7\fR To verify a signature \fI sig.ksig\fR using online verification service call:
.RS
\fBksitool -vx -i\fI sig.ksig\fR
.RE

\fB8\fR To extend a signature \fIsig.ksig\fR and save it as \fIext.ksig\fR call:
.RS
\fBksitool -x -i \fIsig.ksig\fR \fB-o\fI ext.ksig\fR 
.RE

\fB9\fR To verify an extended signature \fI ext.ksig\fR against publication from printed media call:
.RS
\fBksitool -v -i \fIext.ksig\fB --ref\fI AAAAAA-CT5VGY-AAPUCF-L3EKCC-NRSX56-AXIDFL-VZJQK4-WDCPOE-3KIWGB-XGPPM3-O5BIMW-REOVR4 
.RE

\fB10\fR To download a publication file \fIpubfile\fR call:
.RS
\fBksitool -p -o\fI pubfile\fR
.RE

\fB11\fR To verify publication file \fIpubfile\fR call:
.RS
\fBksitool -v -b \fIpubfile\fR
.RE

\fB12\fR To create a publication string call:
.RS
\fBksitool -p -T\fR \fI"2015-10-15 00:00:00"\fR 
.RE


.SH ENVIRONMENT	

. B Default service access URL-s:

To define default URL-s, they must be defined as environment variables. For aggregator and extender service, define environment variables \fB KSI_AGGREGATOR\fR and \fBKSI_EXTENDER\fR with content\fI 'url=<url> pass=<pass> user=<user>'\fR. Only url part is mandatory: user and pass can be left undefined if anonymous access is allowed by the service. Default \fI <pass> \fR and \fI <user> \fRis \fI 'anon'\fR. 
.br

For publications file, define \fBKSI_PUBFILE\fR with content '\fIurl=<url> <constraint> <constraint> \fR...'. Constraint is formatted as  \fI<OID>="<value>"\fR where \fB""\fR can be omitted if 'value' does not contain any white-space characters. Publications file url is mandatory but constraints are not if at least one constraint is defined on command-line (see \fB--cnstr\fR).
.br

Using includes (\fB --inc\fR) or defining urls on command-line will override defaults.

.SH AUTHOR

Guardtime AS, http://www.guardtime.com/


doc/ChangeLog
* BUGFIX: Verification with publication string fixed.
* IMPROVEMNT: Enabled autotools option subdir-objects to remove build errors (on Mac OS and Ubuntu).

Version 1.8

2016-02-15 1.8.1

* BUGFIX: Double free fixed when verifying publications file.

2016-01-26 1.8.0

* IMPROVEMENT: RPM package version 'STABLE' changed to '1'.
* BUGFIX: Curl and Openssl removed from packaging dependencies.
* BUGFIX: When verifying documents hash (-f|-F) signature is verified only once.

2016-01-05 release (1.7.0)
* Requires libksi v3.5 or v3.6 (recommended).
* Parameter -T supports date represented as "YYYY-MM-DD hh:mm:ss" or as seconds since 1970-01-01 00:00:00 UTC.
* When environment variables are defined incorrectly warnings and helpful error messages are printed.
* When C API returns error that publications file URL is not configured, ksitool gives a hint how to do that.
* Bug fix. When downloading publications file it is verified only once.

2015-11-24 release (1.6.0)
* New task dump publications file implemented (-pd -P <url> | [-b <file>]).
* Ksi error trace is not printed with flag -d to stdrerr. Errot trace is pushed to ksi log.
* When verifying signature signer identity (-n) and publication references (-r) are printed with flag -d.
* Windows installer build supports INSTALL_MACHINE as x86, x64, 32 or 64.
* When aggregator url is used for extending or vice versa, helpful error message is printed.
* Debug messages related to PKI signature are improved (use with -d).
* ChangeLog added. 
* Bug fix. When defining publications file constraints on command-line (--cnstr) default constraints are ignored.
* Bug fix. Environment variables with similar prefix to the correct one are ignored.

2015-09-30 release (1.5.0)
* Help text and man page improved.
* Missing parameters value error message improved. "Parameter has no content" replaced with "Parameter must have content".
* New parameter --cnstr oid=value introduced.
* Tcp scheme removed from the list of correct uri schemes and replaced with ksi+tcp.
* Verification of publications file outputs the time of the latest publication in the file.
* Using flag -d KSI error stack is printed to stderr.
* Flag --tlv removed. Use gttlvdump instead.
* Deprecated parameter -E is supported but not documented in help nor man page. It is converted to --cnstr and warning is printed.
* Default values taken from API removed from ksitool.
* New environment variable KSI_PUBFILE implemented for default publications file.
* Bug fix. It is possible to choose libs when linking under Windows.
* Bug fix. Ksitool allows verification of not extended signature using pub. string.
* Bug fix. Debian packages allow libssl0.9.8 or libssl1.0.0.
* Bug fix. Publications file publication printing fixed when reference is missing.

2015-06-12 release (1.3.0a)
* Man page and pdf updated.
* Help updated.
* Error handling refactored.
* Using input file name as '-', it is possible to read data from stdin.
* Using output file name as '-', it is possible to output data to stdout.
* New flags --silent and --nowarn to silent all info and warning or just warning messages.
* --aggre --htime and --setsystime temporarily removed.
* Command-line int type parameters range is [0:INT_MAX].
* Verification with publications string (--ref) implemented.
* New flag --tlv introduced to print signatures tlv structure.
* Bug fix. Parameter -b (publication file) is allowed with -x.

2015-05-15 release (1.3.0)
* Signatures signing time is printed when signing (-s) and verifying (-v) with flag -d.
* Windows packaging supported.
* Windows packaging installs pdf help file.
* DEB packaging implemented.
* Man page for ksitool creaed and packaged.
* License file license.txt added and packaged.
* It is possible to verify against publication string using --ref.

2015-03-31 release (1.2.0)
* Long parameters are defined with -- and short parameters with -.
* Error messages changed.
* Help message changed.
* KSI library and ksitool versions are printed in help.
* Command-line engine is detecting typos and generating corresponding error messages.
* Publication reference format changed.
* Parameters -X and -S can be defined as system variables.
* When overriding service urls on command-line default user and pass is ignored and warning is printed.
* It is possible to extended to a given publication time set as seconds since 1970-01-01 00:00:00 UTC.
* KSI error trace is logged with parameter --log.
* Tcp client supported with url scheme tcp://.
* Url scheme ksi:// is supported.
* Bug fix. Extended signature is actually verified online when using -vx. 
* Bug fix. Service timing measurement fixed.
* Bug fix. When downloading publications file (-p -o -P), parameter -X is not required.
* Bug fix. When extending invalid signature verification result is shown correctly.
* Bug fix. When verifying correctly extended signature, extender url (-X) is not mandatory if correct publications file is available.

2014-12-15 release (1.1.0)
* New parameter -log added to dump KSI library log into file.
* Error messages changed.

2014-11-28 Initial release (1.0.0)
* First release.


src/ksitool.c
/**************************************************************************
 *
 * GUARDTIME CONFIDENTIAL
 *
 * Copyright (C) [2015] Guardtime, Inc
 * All Rights Reserved
 *
 * NOTICE:  All information contained herein is, and remains, the
 * property of Guardtime Inc and its suppliers, if any.
 * The intellectual and technical concepts contained herein are
 * proprietary to Guardtime Inc and its suppliers and may be
 * covered by U.S. and Foreign Patents and patents in process,
 * and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this
 * material is strictly forbidden unless prior written permission
 * is obtained from Guardtime Inc.
 * "Guardtime" and "KSI" are trademarks or registered trademarks of
 * Guardtime Inc.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ksi/compatibility.h>
#include "param_set/param_set.h"
#include "param_set/task_def.h"
#include "tool_box/default_tasks.h"
#include "tool_box/param_control.h"
#include "tool_box/tool_box.h"
#include "tool_box/component.h"
#include "ksitool_err.h"
#include "printer.h"
#include "conf.h"


#ifndef _WIN32
#  ifdef HAVE_CONFIG_H
#    include "config.h"
#  endif
#endif

const char *getVersion(void) {
	static const char versionString[] = VERSION;
	return versionString;
}

const char *getName(void) {
	static const char versionString[] = TOOL_NAME;
	return versionString;
}

static char *hash_algorithms_to_string(char *buf, size_t buf_len) {
	int i;
	size_t count = 0;

	if (buf == NULL || buf_len == 0) {
		return NULL;
	}


	for (i = 0; i < KSI_NUMBER_OF_KNOWN_HASHALGS; i++) {
		if (KSI_isHashAlgorithmSupported(i)) {
			count += KSI_snprintf(buf + count, buf_len - count, "%s%s",
				count == 0 ? "" : " ",
				KSI_getHashAlgorithmName(i)
				);
		}
	}
	if (count > 0) {
		count += KSI_snprintf(buf + count, buf_len - count, ".");
	}

	return buf;
}

static void print_general_help(PARAM_SET *set){
	int res;
	const char *toolVersion = NULL;
	const char *apiVersion = NULL;
	char *aggre_url = NULL;
	char *ext_url = NULL;
	char *pub_url = NULL;
	char *cnstr = NULL;
	char buf[1024];

	if (set == NULL) {
		return;
	}

	toolVersion = getVersion();
	apiVersion = KSI_getVersion();

	res = PARAM_SET_getStr(set, "S", NULL, PST_PRIORITY_HIGHEST, PST_INDEX_LAST, &aggre_url);
	if (res != PST_OK && res != PST_PARAMETER_EMPTY) return;

	res = PARAM_SET_getStr(set, "X", NULL, PST_PRIORITY_HIGHEST, PST_INDEX_LAST, &ext_url);
	if (res != PST_OK && res != PST_PARAMETER_EMPTY) return;

	res = PARAM_SET_getStr(set, "P", NULL, PST_PRIORITY_HIGHEST, PST_INDEX_LAST, &pub_url);
	if (res != PST_OK && res != PST_PARAMETER_EMPTY) return;

	aggre_url = aggre_url != NULL ? aggre_url : "Not defined.";
	ext_url = ext_url != NULL ? ext_url : "Not defined.";
	pub_url = pub_url != NULL ? pub_url : "Not defined.";

	print_info("\nDefault service access URL-s:\n");

	/**
	 * Print information about how to define default service access urls, if at
	 * least one is not defined.
     */
	if (aggre_url == NULL || ext_url == NULL || pub_url == NULL) {
		print_info(
		"  To define default URL-s, system environment variable KSI_CONF must be defined,\n"
		"  that is going to point to ksi configurations file. Configurations file has\n"
		"  similar syntax to the command-line, but parameters are placed line by line.\n"
		"  For extender and aggregator -X <url> [--ext-user <str> --ext-pass <str>] and \n"
		"  -S <url> [--aggr-user <str> --aggr-pass <str>] must be defined. For publications\n"
		"  file -P <url> [--cnstr <oid>=<value>] must be defined.\n\n"
		);
	}

	/**
	 * Print info about default services.
     */
	print_info(
	"  Signing:		%s\n"
	"  Extending/Verifying:	%s\n"
	"  Publications file:	%s\n\n",
		aggre_url, ext_url, pub_url
	);

	/**
	 * Print info about supported hash algorithms.
     */
	print_info(
	"Supported hash algorithms (-H, -F):\n"
	"  %s\n",
		hash_algorithms_to_string(buf, sizeof(buf)));
}

static int ksitool_compo_get(TASK_SET *tasks, PARAM_SET **set, TOOL_COMPONENT_LIST **compo);


int main(int argc, char** argv, char **envp) {
	int res;
	PARAM_SET *set_task_name = NULL;
	PARAM_SET *set = NULL;
	PARAM_SET *configuration = NULL;
	TOOL_COMPONENT_LIST *components = NULL;
	TASK_SET *tasks = NULL;
	TASK *task = NULL;
	int retval = EXIT_SUCCESS;
	char buf[0xffff];

	print_init();

	/**
	 * Define parameter and task set.
     */
	res = PARAM_SET_new("{h}{version}{conf}", &set);
	if (res != PST_OK) goto cleanup;

	res = TASK_SET_new(&tasks);
	if (res != PST_OK) goto cleanup;

	res = CONF_createSet(&configuration);
	if (res != PST_OK) goto cleanup;

	/**
	 * Load the configurations file from environment.
     */
	res = CONF_fromEnvironment(configuration, "KSI_CONF", envp, 0);
	if (res != PST_OK) goto cleanup;


	/**
	 * Get all possible components to run.
     */
	res = ksitool_compo_get(tasks, &set_task_name, &components);
	if (res != PST_OK) {
		printf("Unable to load: Unable to run components");
		goto cleanup;
	}

	/**
	 * Add the values to the set.
     */
	res = PARAM_SET_add(set_task_name, argv[1], NULL, NULL, 0);

	if (argc > 1) {
		PARAM_SET_readFromCMD(argc, argv, set, 0);
	}

	/**
	 * Extract the task.
     */
	res = TASK_SET_analyzeConsistency(tasks, set_task_name, 0.2);
	if (res != PST_OK) goto cleanup;

	res = TASK_SET_getConsistentTask(tasks, &task);
	if (res != PST_OK && res != PST_TASK_ZERO_CONSISTENT_TASKS && res !=PST_TASK_MULTIPLE_CONSISTENT_TASKS) goto cleanup;

	/**
	 * Simple tool help handler.
     */
	if (PARAM_SET_isSetByName(set, "h")) {
		print_info("%s %s (C) Guardtime\n", getName(), getVersion());
		print_info("%s (C) Guardtime\n\n", KSI_getVersion());

		if (task == NULL) {
			print_info("Usage %s <task> <arguments>\n", getName());
			print_info("All known tasks:\n");
			print_info("%s", TOOL_COMPONENT_LIST_toString(components, "  ", buf, sizeof(buf)));
		} else {
			print_info("%s\n", TOOL_COMPONENT_LIST_helpToString(components, TASK_getID(task),buf, sizeof(buf)));
		}

		print_general_help(configuration);

		goto cleanup;
	} else if (PARAM_SET_isSetByName(set, "version")) {
		print_info("%s %s (C) Guardtime\n", getName(), getVersion());
		goto cleanup;
	}

	if (CONF_isInvalid(configuration)) {
		print_errors("KSI Service configuration is invalid:\n");
		print_errors("%s\n", CONF_errorsToString(configuration, "  ", buf, sizeof(buf)));
		res = KT_INVALID_CONF;
		goto cleanup;
	}


	/**
	 * Invalid task. Give user some hints.
     */
	if (task == NULL) {
		print_errors("Error: Invalid task. Read help (-h) or man page.\n");
		if (PARAM_SET_isTypoFailure(set_task_name)) {
			print_info("%s\n", PARAM_SET_typosToString(set_task_name, NULL, buf, sizeof(buf)));
		} else if (PARAM_SET_isUnknown(set_task_name)){
			print_info("%s\n", PARAM_SET_unknownsToString(set_task_name, NULL, buf, sizeof(buf)));
		}

		goto cleanup;
	}


	/**
	 * Run component by its ID.
	 */
	retval = TOOL_COMPONENT_LIST_run(components, TASK_getID(task), argc - 1, argv + 1, envp);
//	printf("PROC RET %X\n", res);
//	printf("Consistent task is: %i\n", TASK_getID(task));
	res = KT_OK;


cleanup:

	if (res != KT_OK) {
		retval = EXIT_FAILURE;
	}

	PARAM_SET_free(set);
	PARAM_SET_free(set_task_name);
	PARAM_SET_free(configuration);
	TASK_SET_free(tasks);
	TOOL_COMPONENT_LIST_free(components);

	return retval;
}

static int ksitool_compo_get(TASK_SET *tasks, PARAM_SET **set, TOOL_COMPONENT_LIST **compo) {
	int res;
	TOOL_COMPONENT_LIST *tmp_compo = NULL;
	PARAM_SET *tmp_set = NULL;

	/**
	 * Create parameter list that contains all known tasks.
     */
	res = PARAM_SET_new("{s|sign}{x|extend}{v|verify}{pubfile}{tlv}{conf}", &tmp_set);
	if (res != PST_OK) goto cleanup;

	res = TOOL_COMPONENT_LIST_new(32, &tmp_compo);
	if (res != PST_OK) goto cleanup;

	/**
	 * Define all components as possible tasks.
     */
	TASK_SET_add(tasks, 0, "Sign", "sign", NULL, NULL, NULL);
	TASK_SET_add(tasks, 1, "Verify", "verify", NULL, NULL, NULL);
	TASK_SET_add(tasks, 2, "extend", "extend", NULL, NULL, NULL);
	TASK_SET_add(tasks, 3, "pubfile", "pubfile", NULL, NULL, NULL);
	TASK_SET_add(tasks, 0xffff, "conf", "conf", NULL, NULL, NULL);

	/**
	 * Define tool component as runnable.
     */
	TOOL_COMPONENT_LIST_add(tmp_compo, "sign", sign_run, sign_help_toString,  sign_get_desc, 0);
	TOOL_COMPONENT_LIST_add(tmp_compo, "verify", verify_run, verify_help_toString, verify_get_desc, 1);
	TOOL_COMPONENT_LIST_add(tmp_compo, "extend", extend_run, extend_help_toString, extend_get_desc, 2);
	TOOL_COMPONENT_LIST_add(tmp_compo, "pubfile", pubfile_run, pubfile_help_toString, pubfile_get_desc, 3);
	TOOL_COMPONENT_LIST_add(tmp_compo, "conf", conf_run, conf_help_toString, conf_get_desc, 0xffff);

	*set = tmp_set;
	*compo = tmp_compo;
	tmp_set = NULL;
	tmp_compo = NULL;
	res = KT_OK;

cleanup:

	PARAM_SET_free(tmp_set);
	TOOL_COMPONENT_LIST_free(tmp_compo);

	return res;
}
